### DISCUSSION TITLE
Need help using jump server to transfer file

### DISCUSSION THREAD

**Discussion by ashishiter** [hierynomus/sshj#859] [Q&A]
Hi Everyone . I am using a jump server with private key and certificate base authentication . The example of the scp command that works looks like 'How can i add a certificate file along with the private key which will be used by the target server . The scp command look like this :
"scp -i id-rsa  -o "ProxyCommand ssh -qx -i id-rsa -o"CertificateFile id-rsa-cert.pub" [svcAuto@jump.mmm.com](mailto:svcAuto@jump.mmm.com) -W %h:%p" foo.bar ec2-user@targetHost:"
I would like to know how to pass  the ProxyCommand.

**Comment by hierynomus** [comment]
```
public class Jump {
    public static void main(String... args)
            throws IOException {
        SSHClient firstHop = new SSHClient();

        firstHop.loadKnownHosts();

        firstHop.connect("localhost");
        try {

            firstHop.authPublickey(System.getProperty("user.name"));

            DirectConnection tunnel = firstHop.newDirectConnection("localhost", 22);

            SSHClient ssh = new SSHClient();
            try {
                ssh.loadKnownHosts();
                ssh.connectVia(tunnel);
                ssh.authPublickey(System.getProperty("user.name"));

                final Session session = ssh.startSession();
                try {
                    final Session.Command cmd = session.exec("ping -c 1 google.com");
                    System.out.println(IOUtils.readFully(cmd.getInputStream()).toString());
                    cmd.join(5, TimeUnit.SECONDS);
                    System.out.println("\n** exit status: " + cmd.getExitStatus());
                } finally {
                    session.close();
                }
            } finally {
                ssh.disconnect();
            }
        } finally {
            firstHop.disconnect();
        }
    }
}
```

i.e. you should open a `DirectConnection` to the jump server, and then to a `sshClient.connectVia(...)` call for connecting through the jump host to your destination host.

**Reply by ashishiter** [reply]
Thanks for the response . Is there a way i can pass the ProxyCommand.
`
'ProxyCommand ssh -qx -i id-rsa -o "CertificateFile id-rsa-cert.pub" svc@jump.com  -W%h:%p' ec2-user@*********** uptime
`
It has -o"CertificateFile id-rsa-cert.pub"  . Not sure where i can pass the certificate file in the tunnel .
The below ssh command work fine for me 
`
ssh -i id-rsa -o 'ProxyCommand ssh -qx -i id-rsa -o "CertificateFile id-rsa-cert.pub" svc@jump.com  -W%h:%p' ec2-user@*********** uptime
`

`

public class SSHJSCPTest {
	   private static final Console con = System.console();
    public static void main(String[] args) throws IOException {
    	
    	Config config = new DefaultConfig();
    	List<Factory.Named<KeyAlgorithm>> algorithms = Arrays.asList(KeyAlgorithms.SSHRSA());
    	
    	config.setKeyAlgorithms(algorithms);
    	
    	SSHClient firstHop = new SSHClient(config);

        firstHop.loadKnownHosts();

        firstHop.connect("jump.svc.com");
        try {

            firstHop.authPublickey("svcAutobahn","/Users/akumar053/new-rsa-keyset/id-rsa");

            DirectConnection tunnel = firstHop.newDirectConnection("ec2-user@********", 22);
            
            
            SSHClient ssh = new SSHClient();
            try {
                ssh.loadKnownHosts();
                ssh.connectVia(tunnel);
                ssh.authPublickey("svcAutobahn","/Users/akumar053/new-rsa-keyset/id-rsa");
                
                

                final Session session = ssh.startSession();
                try {
                    final Session.Command cmd = session.exec("uptime");
                    System.out.println(IOUtils.readFully(cmd.getInputStream()).toString());
                    cmd.join(5, TimeUnit.SECONDS);
                    System.out.println("\n** exit status: " + cmd.getExitStatus());
                } finally {
                    session.close();
                }
            } finally {
                ssh.disconnect();
            }
        } finally {
            firstHop.disconnect();
        }
    	
    	
    }
        
}

`

**Reply by hierynomus** [reply]
There is no need for the ProxyCommand, with the ProxyCommand, the cert file should also already live on your local machine if I'm not mistaken. Also see: https://www.cyberciti.biz/faq/linux-unix-ssh-proxycommand-passing-through-one-host-gateway-server/

**Reply by ashishiter** [reply]
`13:52:59.839 [main] ERROR net.schmizz.concurrent.Promise - <<chan#0 / open>> woke to: Opening `direct-tcpip` channel failed: open failed
13:52:59.839 [sshj-Reader-jump.com/96.118.251.38:22-1683222747014] DEBUG net.schmizz.sshj.connection.ConnectionImpl - Forgetting `direct-tcpip` channel (#0)
13:52:59.840 [sshj-Reader-jump.com/96.118.251.38:22-1683222747014] DEBUG net.schmizz.concurrent.Promise - Setting <<chan#0 / close>> to `SOME`
13:52:59.840 [main] INFO net.schmizz.sshj.transport.TransportImpl - Disconnected - BY_APPLICATION
13:52:59.840 [main] DEBUG net.schmizz.sshj.connection.ConnectionImpl - Notified of net.schmizz.sshj.transport.TransportException: [BY_APPLICATION] Disconnected
13:52:59.840 [main] DEBUG net.schmizz.sshj.transport.TransportImpl - Sending SSH_MSG_DISCONNECT: reason=[BY_APPLICATION], msg=[]
13:52:59.841 [main] DEBUG net.schmizz.concurrent.Promise - Setting <<transport close>> to `SOME`
13:52:59.841 [sshj-Reader-jump.com/96.118.251.38:22-1683222747014] DEBUG net.schmizz.sshj.transport.Reader - Stopping
Exception in thread "main" Opening `direct-tcpip` channel failed: open failed
	at net.schmizz.sshj.connection.channel.direct.AbstractDirectChannel.gotOpenFailure(AbstractDirectChannel.java:74)
	at net.schmizz.sshj.connection.channel.direct.AbstractDirectChannel.gotUnknown(AbstractDirectChannel.java:99)
	at net.schmizz.sshj.connection.channel.AbstractChannel.handle(AbstractChannel.java:204)
	at net.schmizz.sshj.connection.ConnectionImpl.handle(ConnectionImpl.java:130)
	at net.schmizz.sshj.transport.TransportImpl.handle(TransportImpl.java:493)
	at net.schmizz.sshj.transport.Decoder.decode(Decoder.java:113)
	at net.schmizz.sshj.transport.Decoder.received(Decoder.java:200)
	at net.schmizz.sshj.transport.Reader.run(Reader.java:60)
`

**Reply by ashishiter** [reply]
Its all working fine now .. thank for you help.
